Adapt `Storage` and `SnapshotSource` to use `Rc` only.

jayz22 opened the pull request:
### What

Resolves https://github.com/stellar/rs-soroban-env/issues/603. 

This PR explores the 2nd solution described in the issue above. By adapting the `Storage` and `SnapshotSource` to using `Rc`, it eliminates the unnecessary cloning of `LedgerKey` and `LedgerEntry`.

### Why

The `LedgerKey` and `LedgerEntry` are deep structures (contains `ScVal`). Cloning them can be very expensive. Also in order to properly charge the cloning to the budget. `MeteredClone` needs to be implemented for them and all their substructures, which adds bloating to the code. 

### Known limitations

[TODO or N/A]

dmkozh reviewed (commented):
Looks fine to me, though not sure what are the expected additional changes to move this from the 'draft' state to review.

jayz22 commented:
> Looks fine to me, though not sure what are the expected additional changes to move this from the 'draft' state to review.

Thanks @dmkozh. I just want to go through the whole stack: core, sdk, rpc to make sure there are no major disruptions there if this lands. Once that's done, I'll mark this ready for review.

jayz22 commented:
The churn on the downstream (core, sdk, rpc) seems pretty minimal (see the three draft PRs mentioned above), so I'm marking this ready for review.

graydon commented on soroban-env-host/src/host/data_helper.rs:
```diff
@@ -17,69 +18,60 @@ use super::metered_clone::MeteredClone;
-        LedgerKey::ContractCode(LedgerKeyContractCode { hash: wasm_hash })
-    }
-
-    pub(crate) fn retrieve_contract_code_from_storage(
```
Why was this helper deleted?

jayz22 replied:
Added it back. See comment above.

graydon commented on soroban-env-host/src/host/data_helper.rs:
```diff
@@ -17,69 +18,60 @@ use super::metered_clone::MeteredClone;
         contract_source: ScContractCode,
         contract_id: Hash,
-        key: &LedgerKey,
+        key: Rc<LedgerKey>,
```
This can be `&Rc<...>`

jayz22 replied:
Fixed.

graydon reviewed (changes requested):
Solid step in the right direction, but a few things to clean up.

jayz22 commented:
@graydon I've addressed all your feedback around changing `Rc->&Rc` at the interface level. It makes a lot more sense:
1. Reduces cloning. `Rc::clone` is only called at the lowest level when it has to, avoiding cloning it everywhere upfront (where most of the lower code paths don't actually need it). 
2. Makes the interface more uniform and much easier to use.

jayz22 commented on soroban-env-host/src/storage.rs:
```diff
@@ -71,39 +70,35 @@ pub struct Footprint(pub FootprintMap);
     pub fn record_access(
         &mut self,
-        key: &LedgerKey,
+        key: Rc<LedgerKey>,
```
Thank you, fixed!

jayz22 commented on soroban-env-host/src/storage.rs:
```diff
@@ -185,52 +180,50 @@ impl Storage {
-    pub fn get(&mut self, key: &LedgerKey, budget: &Budget) -> Result<LedgerEntry, HostError> {
+    pub fn get(
+        &mut self,
+        key: Rc<LedgerKey>,
```
Fixed.

jayz22 commented on soroban-env-host/src/storage.rs:
```diff
@@ -185,52 +180,50 @@ impl Storage {
         &mut self,
-        key: &LedgerKey,
-        val: Option<LedgerEntry>,
+        key: Rc<LedgerKey>,
```
Fixed.

jayz22 commented on soroban-env-host/src/storage.rs:
```diff
@@ -245,11 +238,11 @@ impl Storage {
         &mut self,
-        key: &LedgerKey,
-        val: &LedgerEntry,
+        key: Rc<LedgerKey>,
```
Fixed.

jayz22 commented on soroban-env-host/src/storage.rs:
```diff
@@ -261,7 +254,7 @@ impl Storage {
     /// [LedgerKey] has been declared in the [Footprint] as
     /// [AccessType::ReadWrite].
-    pub fn del(&mut self, key: &LedgerKey, budget: &Budget) -> Result<(), HostError> {
+    pub fn del(&mut self, key: Rc<LedgerKey>, budget: &Budget) -> Result<(), HostError> {
```
Fixed.

jayz22 commented on soroban-env-host/src/storage.rs:
```diff
@@ -274,22 +267,22 @@ impl Storage {
     /// In [FootprintMode::Enforcing] mode, succeeds only if the access has been
     /// declared in the [Footprint].
-    pub fn has(&mut self, key: &LedgerKey, budget: &Budget) -> Result<bool, HostError> {
+    pub fn has(&mut self, key: Rc<LedgerKey>, budget: &Budget) -> Result<bool, HostError> {
```
Fixed.

jayz22 commented on soroban-env-host/src/host.rs:
```diff
@@ -842,11 +840,18 @@ impl Host {
             #[cfg(feature = "vm")]
             ScContractCode::WasmRef(wasm_hash) => {
-                let code_entry = self.retrieve_contract_code_from_storage(wasm_hash)?;
+                let key = self.contract_code_ledger_key(&wasm_hash)?;
```
I initially removed this helper because it was a little hard to refactor it with the change of interface to `Rc` (borrow checker related). But I realize this is probably not a good reason for removing it.
With the `Rc->&Rc` fix up, this makes it a lot easier to work with. 
So I have added it back.

jayz22 commented on soroban-env-host/src/host.rs:
```diff
@@ -1053,10 +1058,10 @@ impl Host {
         &self,
-        key: LedgerKey,
-        val: soroban_env_common::xdr::LedgerEntry,
+        key: Rc<LedgerKey>,
```
Fixed.
